File: RssReader/RssReader/RssReader/ImageCache/ImageCache.m
/*!
 *@file		ImageCache.m
 *@brief	ImageCache実装ファイル
 *@author	Satoshi Iwaki
 *@date		2013/05/16
 */

#import "ImageCache.h"

@implementation ImageCache
{
    NSCache *cache;
    NSOperationQueue *operationQueue;
    UIImage *defaultImage;
    NSURLSession *session;
}

@synthesize defaultImage;

//*********** Singleton pattern ***********

static id _sharedInstance = nil;

+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [[self alloc] init];
    });
    return _sharedInstance;
}

+ (void)deleteInstance
{
    _sharedInstance = nil;
}

- (id)copyWithZone:(NSZone *)zone
{
    return self;
}

//*********** Singleton pattern ***********

/*!
 *@brief		初期化処理
 *@return       シングルトンインスタンス
 */
- (instancetype)init
{
    self = [super init];
    if (self) {
        cache = [[NSCache alloc] init];
        operationQueue = [[NSOperationQueue alloc] init];
        [operationQueue setMaxConcurrentOperationCount:1];
        defaultImage = [[UIImage imageNamed:@"NoImage"] stretchableImageWithLeftCapWidth:16 topCapHeight:16];
        session = [NSURLSession sessionWithConfiguration:NSURLSessionConfiguration.defaultSessionConfiguration];

    }
    return self;
}

/*!
 *@brief		メモリ解放処理
 */
- (void)dealloc
{
    [session invalidateAndCancel];
    cache = nil;
    [operationQueue cancelAllOperations];
    operationQueue = nil;
    defaultImage = nil;
}


#pragma mark - Public methods

- (void)clearQueue
{
    [operationQueue cancelAllOperations];
}

- (void)clearCache
{
    [cache removeAllObjects];
}

- (UIImage *)cachedImageForURL:(NSURL *)URL
{
    if (URL == nil) {
        return nil;
    }
    
    return [cache objectForKey:URL];
}

- (void)getImageForURL:(NSURL *)URL completionHandler:(void (^)(UIImage *image))completionHandler
{
    if (URL == nil) {
        return;
    }
    
    __block UIImage *image = [cache objectForKey:URL];
    
    if (image) {
        completionHandler(image);
    } else {
        NSOperation *operation = [NSBlockOperation blockOperationWithBlock:^{
            image = [cache objectForKey:URL];
            
            if (image) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    completionHandler(image);
                });
            } else {
                NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:URL
                                                                            cachePolicy:NSURLRequestReloadIgnoringLocalCacheData
                                                                        timeoutInterval:60.0];
                
                // HTTP通信開始
                NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
                    if (error) {
                        image = [self defaultImage];
                    } else {
                        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
                        NSInteger statusCode = [httpResponse statusCode];
                        
                        if (200 == statusCode) {
                            if (data) {
                                image = [UIImage imageWithData:data];
                            } else {
                                // データが空の場合
                                image = defaultImage;
                            }
                        } else {
                            // ステータスコードが正常(200)以外の場合
                            image = defaultImage;
                        }
                    }
                    
                    if (image) {
                        [cache setObject:image forKey:URL];
                    }
                    
                    dispatch_async(dispatch_get_main_queue(), ^{
                        completionHandler(image);
                    });
                }];
                [task resume];
            }
        }];
        [operationQueue addOperation:operation];
    }
}

- (void)imageForURL:(NSURL *)URL target:(id)target selector:(SEL)aSelector
{
    if (URL == nil) {
        return;
    }
    
    __block UIImage *image = [cache objectForKey:URL];
    
    if (image) {
        if ([target respondsToSelector:aSelector]) {
            [target performSelector:aSelector
                         withObject:image];
        }
    } else {
        NSOperation *operation = [NSBlockOperation blockOperationWithBlock:^{
            image = [cache objectForKey:URL];
            
            if (image) {
                [target performSelectorOnMainThread:aSelector
                                         withObject:image
                                      waitUntilDone:NO];
            } else {
                NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:URL
                                                                            cachePolicy:NSURLRequestReloadIgnoringLocalCacheData
                                                                        timeoutInterval:60.0];
                
                // HTTP通信開始
                NSURLSessionDataTask *task = [session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
                    if (error) {
                        image = [self defaultImage];
                    } else {
                        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
                        NSInteger statusCode = [httpResponse statusCode];
                        
                        if (200 == statusCode) {
                            if (data) {
                                image = [UIImage imageWithData:data];
                            } else {
                                // データが空の場合
                                image = defaultImage;
                            }
                        } else {
                            // ステータスコードが正常(200)以外の場合
                            image = defaultImage;
                        }
                    }
                    
                    if (image) {
                        [cache setObject:image forKey:URL];
                    }
                    
                    if ([target respondsToSelector:aSelector]) {
                        [target performSelectorOnMainThread:aSelector
                                                 withObject:image
                                              waitUntilDone:NO];
                    }
                }];
                [task resume];
            }
        }];
        [operationQueue addOperation:operation];
    }
}

+ (UIImage *)defaultImage
{
    return [[ImageCache sharedInstance] defaultImage];
}

@end

File: RssReader/RssReader/RssReader/ImageCache.swift
//
//  ImageCache.swift
//  RssReader
//
//  Created by Iwaki Satoshi on 2017/08/02.
//  Copyright © 2017年 Satoshi Iwaki. All rights reserved.
//

import UIKit

class ImageCache: NSObject, NSCacheDelegate {
    static let sharedInstance = ImageCache()
    
    private let cache = NSCache<NSString, UIImage>()
    private let defaultImage = UIImage(named: "NoImage")!
    private let session = URLSession(configuration: URLSessionConfiguration.default, delegate: nil, delegateQueue:OperationQueue.main)
    private let queue = OperationQueue()
    
    override init() {
        super.init()
        cache.countLimit = 50
        cache.delegate = self
        queue.maxConcurrentOperationCount = 1
    }
    
    deinit {
        session.invalidateAndCancel()
        queue.cancelAllOperations()
        queue.waitUntilAllOperationsAreFinished()
    }
    
    func cachedImage(url: URL) -> UIImage? {
        guard let image = cache.object(forKey: url.absoluteString as NSString) else {
            return nil
        }
        return image
    }
    
    func getImage(url: URL, completionHandler: @escaping (UIImage?) -> Void) {
        queue.addOperation {
            if let image = self.cachedImage(url: url) {
                DispatchQueue.main.async {
                    completionHandler(image)
                }
                return
            }
            let semaphore = DispatchSemaphore(value: 0)
            DispatchQueue.global().async {
                let request = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 60.0)
                self.session.dataTask(with: request) { (data, response, error) in
                    if let error = error {
                        debugPrint("error = \(error)");
                        completionHandler(nil)
                        semaphore.signal()
                        return
                    }
                    guard let httpResponse = response as? HTTPURLResponse, let data = data else {
                        debugPrint("HTTP response is nil or data is nil")
                        completionHandler(nil)
                        semaphore.signal()
                        return
                    }
                    guard httpResponse.statusCode == 200 else {
                        debugPrint("HTTP response code is not 200")
                        completionHandler(nil)
                        semaphore.signal()
                        return
                    }
                    guard let image = UIImage(data: data) else {
                        debugPrint("HTTP response data is not image")
                        completionHandler(nil)
                        semaphore.signal()
                        return
                    }
                    self.cache.setObject(image, forKey: url.absoluteString as NSString)
                    completionHandler(image)
                    semaphore.signal()
                }.resume()
            }
            semaphore.wait()
        }
    }
    
    func clearCache() {
        cache.removeAllObjects()
    }
}

extension UIImageView {
    func setImage(url: URL) {
        if let image = ImageCache.sharedInstance.cachedImage(url: url) {
            self.image = image
            return
        }
        
        ImageCache.sharedInstance.getImage(url: url) { (image: UIImage?) in
            guard let image = image else {
                self.image = UIImage(named:"NoImage.png")
                return
            }
            
            self.image = image
            self.setNeedsLayout()
            self.setNeedsDisplay()
        }
    }
}

